pkg/trie/triedb/node.go

```diff
@@ -0,0 +1,195 @@
+// Copyright 2024 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package triedb
+
+import (
+	"bytes"
+
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/pkg/trie/db"
+	"github.com/ChainSafe/gossamer/pkg/trie/triedb/codec"
+)
+
+type StorageHandle struct{ int }
+
+func (sh StorageHandle) toNodeHandle() NodeHandle {
+	return InMemory{idx: sh}
+}
+
+type NodeHandle interface {
+	isNodeHandle()
+}
+
+type (
+	InMemory struct {
+		idx StorageHandle
+	}
+	Hash struct {
+		hash common.Hash
+	}
+)
+
+func (InMemory) isNodeHandle() {}
+func (Hash) isNodeHandle()     {}
+
+func newFromEncodedMerkleValue(
+	parentHash common.Hash,
+	encodedNodeHandle codec.MerkleValue,
+	storage NodeStorage,
+) (NodeHandle, error) {
+	switch encoded := encodedNodeHandle.(type) {
+	case codec.HashedNode:
+		return Hash{hash: common.NewHash(encoded.Data)}, nil
+	case codec.InlineNode:
+		child, err := newNodeFromEncoded(parentHash, encoded.Data, storage)
+		if err != nil {
+			return nil, err
+		}
+		return InMemory{storage.alloc(New{child})}, nil
+	default:
+		panic("unreachable")
+	}
+}
+
+type Value interface {
+	isValue()
+	getHash() common.Hash
+}
+
+type (
+	Inline struct {
+		Data []byte
+	}
+
+	ValueRef struct {
+		hash common.Hash
+	}
+
+	NewValueRef struct {
+		hash *common.Hash
+		Data []byte
+	}
+)
+
+func (Inline) isValue()                  {}
+func (Inline) getHash() common.Hash      { return common.EmptyHash }
+func (ValueRef) isValue()                {}
+func (vr ValueRef) getHash() common.Hash { return vr.hash }
+func (NewValueRef) isValue()             {}
+func (vr NewValueRef) getHash() common.Hash {
+	if vr.hash == nil {
+		return common.EmptyHash
+	}
+
+	return *vr.hash
+}
+
+func NewValue(data []byte, threshold int) Value {
+	if len(data) >= threshold {
+		return NewValueRef{Data: data}
+	}
+
+	return Inline{Data: data}
+}
+
+func NewFromEncoded(encodedValue codec.NodeValue) Value {
+	switch encoded := encodedValue.(type) {
+	case codec.InlineValue:
+		return Inline{Data: encoded.Data}
+	case codec.HashedValue:
+		return ValueRef{hash: common.NewHash(encoded.Data)}
+	}
+
+	return nil
+}
+
+func InMemoryFetchedValue(value Value, prefix []byte, db db.DBGetter, fullKey []byte) ([]byte, error) {
+	switch v := value.(type) {
+	case Inline:
+		return v.Data, nil
+	case NewValueRef:
```

Comment:
Could you elaborate on the different between the types `NewValueRef` and `ValueRef`?

Reply:
Yes I just replied about it in a different [comment](https://github.com/ChainSafe/gossamer/pull/3954/files/7abec593d534e923da5f02f17c34cbc7c90eb11b#r1610619630) 😄